internal/webhook/v1alpha1/workspace_webhook.go

```diff
@@ -115,14 +131,59 @@ func (v *WorkspaceCustomValidator) ValidateCreate(_ context.Context, obj runtime
 }
 
 // ValidateUpdate implements webhook.CustomValidator so a webhook will be registered for the type Workspace.
-func (v *WorkspaceCustomValidator) ValidateUpdate(_ context.Context, oldObj, newObj runtime.Object) (admission.Warnings, error) {
-	workspace, ok := newObj.(*workspacesv1alpha1.Workspace)
+func (v *WorkspaceCustomValidator) ValidateUpdate(ctx context.Context, oldObj, newObj runtime.Object) (admission.Warnings, error) {
+	oldWorkspace, ok := oldObj.(*workspacesv1alpha1.Workspace)
+	if !ok {
+		return nil, fmt.Errorf("expected a Workspace object for the oldObj but got %T", oldObj)
+	}
+	newWorkspace, ok := newObj.(*workspacesv1alpha1.Workspace)
 	if !ok {
 		return nil, fmt.Errorf("expected a Workspace object for the newObj but got %T", newObj)
 	}
-	workspacelog.Info("Validation for Workspace upon update", "name", workspace.GetName())
 
-	// TODO(user): fill in your validation logic upon object update.
+	// Validate that ownership annotations are immutable
+	if oldWorkspace.Annotations != nil && newWorkspace.Annotations != nil {
+		if oldCreatedBy := oldWorkspace.Annotations[controller.AnnotationCreatedBy]; oldCreatedBy != "" {
+			if newCreatedBy := newWorkspace.Annotations[controller.AnnotationCreatedBy]; newCreatedBy != oldCreatedBy {
+				return nil, fmt.Errorf("created-by annotation is immutable")
+			}
+		}
+		if oldLastUpdatedBy := oldWorkspace.Annotations[controller.AnnotationLastUpdatedBy]; oldLastUpdatedBy != "" {
+			if newLastUpdatedBy := newWorkspace.Annotations[controller.AnnotationLastUpdatedBy]; newLastUpdatedBy != oldLastUpdatedBy {
+				return nil, fmt.Errorf("last-updated-by annotation is immutable")
+			}
+		}
+	}
+
+	// Check accessType transition rules first
+	if oldWorkspace.Spec.AccessType == accessTypePublic && newWorkspace.Spec.AccessType == accessTypeOwnerOnly {
```

Comment:
are you sure? I think that's one of the user story (see "Upgrade a private space")